### PR TITLE
feat: add telemetry commands

### PR DESCRIPTION
**Pull request by emily-shen**
Fixes [DEVX-1480](https://jira.cfdata.org/browse/DEVX-1480)

Adds new `wrangler telemetry disable` and `wrangler telemetry enable` commands which globally toggle telemetry collection across all projects on that machine.

Adds a new `wrangler telemetry status` command which shows the current telemetry setting, resolved to the current project.

With send_metrics set in wrangler.toml
Screenshot 2024-11-15 at 09 44 59

With WRANGLER_SEND_METRICS environment variable set
Screenshot 2024-11-15 at 09 45 20

Without either send_metrics or WRANGLER_SEND_METRICS set
Screenshot 2024-11-15 at 09 44 18

We prioritise env var > wrangler.toml > global setting via command.

NB we're using a feature branch so will add one changeset when we merge that in.

---

- Tests
  - [ ] TODO (before merge)
  - [x] Tests included
  - [ ] Tests not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [ ] Required
  - [x] Not required because:  no existing e2e's, adequately covered by unit tests
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Documentation not necessary because: currently hidden. Will add later.

### CODE REVIEW

File: packages/wrangler/src/metrics/commands.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	command: "wrangler telemetry status",
+	metadata: {
+		description: "Check whether Wrangler telemetry collection is enabled",
+		owner: "Workers: Authoring and Testing",
+		status: "stable",
+	},
+	async handler(_, { config }) {
+		const metricsConfig = await getMetricsConfig({
+			sendMetrics: config.send_metrics,
+		});
```
Comment (edmundhung) [inline packages/wrangler/src/metrics/commands.ts:60]: Would `readMetricsConfig` be enough? `getMetricsConfig` does more to get userId and will re-confirm whether users wanna enable telemetry in some cases which seems unnecessary here.

File: .changeset/little-crabs-rule.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+"wrangler": minor
+---
+
+feat: add new `wrangler telemetry` commands to configure whether Wrangler collects usage telemetry
```
Comment (edmundhung) [inline .changeset/little-crabs-rule.md:5]: Should we make it clear that these commands are currently hidden and will be exposed in a future release once telemetry collection is fully implemented?

File: packages/wrangler/src/metrics/commands.ts (hunk 1 above)
Comment (emily-shen) [reply packages/wrangler/src/metrics/commands.ts:60]: oh yeah meant to flag this - should `status` be per-project? i.e. should it take into account send_metrics in wrangler.toml at all. If we strictly want to to be about user-level, we should probably clarify that in the messaging. And also we should probably clarify the order of precedence for wrangler.toml vs env variable vs set-by-command.

Also it does re-confirm currently, but I figure we would remove that when we switch to on-by-default.

File: .changeset/little-crabs-rule.md (hunk 2 above)
Comment (emily-shen) [reply .changeset/little-crabs-rule.md:5]: yep! also wondering if this is 'experimental' and thus a patch?

Comment (edmundhung) [reply .changeset/little-crabs-rule.md:5]: Either way works for me 🤔

File: packages/wrangler/src/metrics/commands.ts (hunk 1 above)
Comment (edmundhung) [reply packages/wrangler/src/metrics/commands.ts:60]: I would expect the status to show the actual stored preference—basically, what the telemetry command is modifying. But yeah, we should make sure the message is clear here.

Comment (edmundhung) [reply packages/wrangler/src/metrics/commands.ts:60]: I feel like we should show two status: Global / Project (which reflects the send_metrics config?)

File: .changeset/little-crabs-rule.md (hunk 2 above)
Comment (emily-shen) [reply .changeset/little-crabs-rule.md]: d697b4da81c39e31e6d1379950fb32876bc3e895

File: packages/wrangler/src/__tests__/metrics.test.ts [hunk 3]
```diff
@@ -434,6 +435,98 @@ describe("metrics", () => {
...
+			await runWrangler("telemetry disable");
+			expect(std.out).toMatchInlineSnapshot(`
+				"Status: Disabled
+
+				Wrangler is no longer collecting telemetry about your usage.
+				"
+			`);
+			expect(await getMetricsConfig({})).toMatchObject({
+				enabled: false,
+			});
```
[the remark this answers is not in the document]
Comment (emily-shen) [reply packages/wrangler/src/__tests__/metrics.test.ts]: 9bdd400be8d809750b366a8a11ada603f09ad423

File: packages/wrangler/src/metrics/commands.ts (hunk 1 above)
Comment (emily-shen) [reply packages/wrangler/src/metrics/commands.ts]: what do you think about this - b44e1b5e3f7e67f54a826a502dc550c6a369991d 
And do you think we should explain how global/project is set underneath, or is just the status fine?

Comment (edmundhung) [reply packages/wrangler/src/metrics/commands.ts]: Let's get some feedbacks from Nevi and Liz on this.

Comment (emily-shen) [reply packages/wrangler/src/metrics/commands.ts]: resolved in 7bc5de1fa7707d6731d8e82d1b417f793f9fcc67 after discussions with Nevi/Liz

File: packages/wrangler/src/__tests__/metrics.test.ts [hunk 4]
```diff
@@ -434,6 +392,191 @@ describe("metrics", () => {
...
+					},
+				});
+				await runWrangler("telemetry status");
+				expect(std.out).toMatchInlineSnapshot(`
+					"Status: Enabled
+
+					To configure telemetry globally on this machine, you can run \`wrangler telemetry disable / enable\`.
+					You can override this for individual projects with the environment variable \`WRANGLER_SEND_METRICS=true/false\`.
+					"
+				`);
```
Comment (edmundhung) [inline packages/wrangler/src/__tests__/metrics.test.ts:419]: I think we can skip this assertion since we will have a full snapshot next anyway.

**Review by edmundhung — APPROVED** [review]
LGTM

**Comment by edmundhung** [comment]
For posterity, there is no changeset as this is merged only into a feature branch.

**Pull request merged**